x-pack/plugin/wildcard/src/main/java/org/elasticsearch/xpack/wildcard/mapper/WildcardFieldMapper.java

```diff
@@ -216,218 +231,383 @@ public WildcardFieldType clone() {
+                @Override
+                public String normalize(String token) {
+                    return toLowerCase(token);
                 }
-
-                // Add any start/end of string character
-                if (i == 0 && patternStructure.openStart == false) {
-                    // Start-of-string anchored (is not a leading wildcard)
-                    fragment = TOKEN_START_OR_END_CHAR + fragment;
+            });
+            Query approxNgramQuery = rewriteBoolToNgramQuery(approxBooleanQuery);
+            
+            // MatchAll is a special case meaning the regex is known to match everything .* and 
+            // there is no need for verification.
+            if (approxNgramQuery instanceof MatchAllDocsQuery) {
+                return new DocValuesFieldExistsQuery(name());
+            }
+            AutomatonQueryOnBinaryDv verifyingQuery = new AutomatonQueryOnBinaryDv(name(), value, automaton);
+            
+            // MatchAllButRequireVerificationQuery is a special case meaning the regex is reduced to a single
+            // clause which we can't accelerate at all and needs verification. Example would be ".." 
+            if (approxNgramQuery instanceof MatchAllButRequireVerificationQuery) {
+                return verifyingQuery;
+            }
+            
+            if (approxNgramQuery != null) {
+                // We can accelerate execution with the ngram query
+                BooleanQuery.Builder verifyingBuilder = new BooleanQuery.Builder();
+                verifyingBuilder.add(new BooleanClause(approxNgramQuery, Occur.MUST));
+                verifyingBuilder.add(new BooleanClause(verifyingQuery, Occur.MUST));
+                return verifyingBuilder.build();
+            }
+            return verifyingQuery;                        
+        }
+        
+        private static String toLowerCase(String string) {
+            return lowercaseNormalizer.normalize(null, string).utf8ToString();
+        }
+        
+        // Takes a BooleanQuery + TermQuery tree representing query logic and rewrites using ngrams of appropriate size.
+        private Query rewriteBoolToNgramQuery(Query approxQuery) {
+            //TODO optimise more intelligently so we: 
+            // 1) favour full-length term queries eg abc over short eg a* when pruning too many clauses.
+            // 2) make MAX_CLAUSES_IN_APPROXIMATION_QUERY a global cap rather than per-boolean clause.
+            if (approxQuery == null) {
+                return null;
+            }
+            if (approxQuery instanceof BooleanQuery) {
+                BooleanQuery bq = (BooleanQuery) approxQuery;
+                BooleanQuery.Builder rewritten = new BooleanQuery.Builder();
+                int clauseCount = 0;
+                for (BooleanClause clause : bq) {
+                    Query q = rewriteBoolToNgramQuery(clause.getQuery());
+                    if (q != null) {
```

Comment:
When are we expecting null queries ? Should we also ignore `MatchAllQuery` since it's a valid query to return from `rewriteBoolToNgramQuery` ?

Reply:
I think we can lose the null check.
MatchAllDocsQuery is intended to indicate that we have a `.*` type regex that needs no verification (unlike a `..` type regex where a MatchAllButRequireVerificationQuery indicates we need to just run the verification query).